src/mozanalysis/experiments.py

```diff
@@ -1,165 +1,268 @@
+                F.expr(
+                    "SUM(sum_active_hours) "
+                    "/ ({hours_epsilon} + SUM(sum_total_hours))".format(**CONSTANTS)
+                ).alias("engagement_intensity"),
+            )
+        )
 
-        data = []
-        branches = self.get_split_by_values(dataset)
-
-        for m in self._metrics:
-            # TODO: Use a lib or make this more robust.
-            metric_name = m.name.replace(" ", "_").lower()
-
-            agg_df = (
-                dataset.select(*([self._aggregate_by, self._split_by] + m.columns))
-                .groupBy(self._aggregate_by, self._split_by)
-                .agg(*m.aggregations)
-                .select(
-                    *[
-                        self._aggregate_by,
-                        self._split_by,
-                        m.final_expression.alias(metric_name),
-                    ]
-                )
+        return agg_df
+
+    def retention(self, week=3):
+        # Compute only the retention metrics.
+        # Return pandas dataframe with data and CIs.
+        if not self._df:
+            self._df = self.get_data(metric=CoreMetrics.RETENTION)
+
+        retention_df = (
+            self._df.withColumn(
+                "enrollment_date",
+                F.min("submission_date").over(Window().partitionBy("client_id")),
+            )
+            .withColumn(
+                "sum_total_hours",
+                self._df.subsession_length / CONSTANTS["seconds_per_hour"],
+            )
+            .withColumn(
+                "sum_total_uris",
+                self._df.scalar_parent_browser_engagement_total_uri_count,
+            )
+            .select(
+                "enrollment_date",
+                "sum_total_hours",
+                "sum_total_uris",
+                F.datediff("submission_date", "enrollment_date").alias("nd"),
+                F.floor(F.datediff("submission_date", "enrollment_date") / 7).alias(
+                    "week"
+                ),
+                *self._df.columns
+            )
+            .filter(F.col("nd") >= 0)
+        )
+
+        week3_df = (
+            retention_df.filter("week={week}".format(week=week))
+            .groupBy(*AGG_COLS)
+            .agg(
+                F.expr("CASE WHEN SUM(sum_total_hours) > 0 THEN 1 ELSE 0 END").alias(
+                    "retention"
+                ),
+                F.expr("MAX(CASE WHEN sum_total_uris >= 5 THEN 1 ELSE 0 END)").alias(
+                    "active_retention"
+                ),
             )
+        )
+
+        return week3_df
+
+    def analyze(self):
+        # TODO
+        # Compute all metrics.
+        #
+        # This is a convenience method.
+        #
+        # The idea would be to make the calls to `engagement()` and `retention`
+        # separately, pass those results to `bootstrap`, the concatenate the
+        # pandas dataframes returned from bootstrap and return that.
```

Comment:
@tdsmith This is one of my questions... whether this would be convenient and useful or not?

Reply:
Maybe a nice-to-have; I don't think it's essential.